mars/services/task/execution/ray/executor.py

```diff
@@ -0,0 +1,265 @@
+    for chunk in chunk_graph.topological_iter():
+        if chunk.key not in context:
+            execute(context, chunk.op)
+
+    output = {key: data for key, data, _ in iter_output_data(chunk_graph, context)}
+    output_meta = {}
+    for chunk in chunk_graph.result_chunks:
+        if chunk.key in output_meta_keys:
+            if isinstance(chunk.op, Fuse):
+                # fuse op
+                chunk = chunk.chunk
+            output_meta[chunk.key] = get_chunk_params(chunk)
+    assert len(output_meta_keys) == len(output_meta)
+
+    logger.info("Finish executing stage: %s", stage_id)
+    has_meta = bool(output_meta_keys)
+    if len(output) + has_meta == 1:
+        return next(iter(output.values()))
+    else:
+        output_values = list(output.values())
+        return [output_meta] + output_values if has_meta else output_values
+
+
+@register_executor_cls
+class RayTaskExecutor(TaskExecutor):
+    name = "ray"
+
+    def __init__(
+        self,
+        config,
+        task,
+        tile_context,
+        ray_executor,
+        cluster_api,
+        lifecycle_api,
+        meta_api,
+    ):
+        self._config = config
+        self._task = task
+        self._tile_context = tile_context
+        self._ray_executor = ray_executor
+
+        # api
+        self._cluster_api = cluster_api
+        self._lifecycle_api = lifecycle_api
+        self._meta_api = meta_api
+
+        self._task_context = {}
+
+    @classmethod
+    async def create(
+        cls,
+        config: Dict,
+        *,
+        session_id: str,
+        address: str,
+        task,
+        tile_context,
+        **kwargs
+    ) -> "TaskExecutor":
+        ray_executor = ray.remote(execute_subtask_chunk_graph)
+        cluster_api, lifecycle_api, meta_api = await cls._get_apis(session_id, address)
+        return cls(
+            config,
+            task,
+            tile_context,
+            ray_executor,
+            cluster_api,
+            lifecycle_api,
+            meta_api,
+        )
+
+    @classmethod
+    @alru_cache(cache_exceptions=False)
+    async def _get_apis(cls, session_id: str, address: str):
+        # TODO(fyrestone): Remove ClusterAPI usage.
+        return await asyncio.gather(
+            ClusterAPI.create(address),
+            LifecycleAPI.create(session_id, address),
+            MetaAPI.create(session_id, address),
+        )
+
+    async def execute_subtask_graph(
+        self,
+        stage_id: str,
+        subtask_graph: SubtaskGraph,
+        chunk_graph: ChunkGraph,
+        tile_context: Dict[TileableType, TileableType],
+        context: Any = None,
+    ) -> Dict[Chunk, ExecutionChunkResult]:
+        logger.info("Stage %s start.", stage_id)
+        context = self._task_context  # TODO(fyrestone): Load context from meta service.
+        output_meta_object_refs = []
+
+        key_to_result_chunks = {}
+        for chunk in chunk_graph.result_chunks:
+            if isinstance(chunk.op, Fuse):
```

Comment:
@qinxuye Do we still has `Fuse` node after coloring merged?

Reply:
There is no `Fuse` op now, it's only used in worker when generating numexpr optimized op.

Reply:
I'm not sure whether we need the `isinstance(chunk.op, Fuse)` check here.